test/IRGen/moveonly_enum_deinits.swift

```diff
@@ -1,9 +1,49 @@
-// RUN: %target-swift-emit-irgen \
+// RUN: %target-swift-emit-irgen                             \
 // RUN:     -enable-experimental-feature NoncopyableGenerics \
-// RUN:     %s \
-// RUN: | \
+// RUN:     -disable-type-layout                             \
+// RUN:     %s                                               \
+// RUN: |                                                    \
 // RUN: %FileCheck %s
 
+struct FIFO<T>: ~Copyable {
+    private var data: UnsafeManualArray<T>
+
+    private(set) public var count: Int
+
+    private var head: Int
+
+    init(withCapacity: Int) {
+        self.head = 0
+        self.count = 0
+        self.data = .init(capacity: withCapacity)
+    }
+
+    deinit {
+    }
+}
+
+// Check that the outlined retain function has a parameter for the type's
+// generic parameter and that it's passed along to the deinit.
+// CHECK-LABEL: define{{.*}} ptr @"$s21moveonly_enum_deinits17UnsafeManualArrayVyxGlWOs"(
+// CHECK-SAME:      ptr %0, 
+// CHECK-SAME:      ptr %T)
+// CHECK-SAME:  {
+// CHECK:         call swiftcc void @"$s21moveonly_enum_deinits17UnsafeManualArrayVfD"(
+// CHECK-SAME:        ptr %T)
+// CHECK:       }
```

Comment:
There's no point in an outlined release that just calls a `deinit` that's already a separate function. How hard would it be to not outline these value witness operations at all for a noncopyable type that directly has a `deinit`?

Reply:
That's a nice peephole.  Implemented here: https://github.com/apple/swift/pull/71787 .  There's a similar one to do for the destroy.

Unfortunately, it's not always applicable.  Generic move-only structs _without_ deinits that contain generic move-only structs _with_ deinits should still get outlined value functions for release and destroy and those outlined value functions need the metadata to pass to the deinits of the contained move-only structs with deinits.  That's implemented here: https://github.com/apple/swift/pull/71799